cmd/ndt7-client/main.go

```diff
@@ -0,0 +1,154 @@
+//
+// The `-hostname <hostname>` flag specifies the hostname to use for
+// performing the ndt7 test. The default is to auto-discover a suitable
+// server by using Measurement Lab's locate service.
+//
+// The `-timeout <timeout>` flag specifies after how many seconds a
+// running ndt7 test should timeout. The default is a large enough
+// value that should be suitable for common conditions.
+//
+// Additionally, passing any unrecognized flag, such as `-help`, will
+// cause ndt7-client to print a brief help message.
+//
+// Event emitted in batch mode
+//
+// This section describes the events emitted in batch mode. The code
+// will always emit a single event per line. In some cases we have
+// wrapped long event lines, below, to simplify reading.
+//
+// When the download subtest starts, this event is emitted:
+//
+//   {"key":"status.measurement_start","value":{"subtest":"download"}}
+//
+// After this event is emitted, we discover the server to use (unless it
+// has been configured by the user) and we connect to it. If any of these
+// operations fail, this event is emitted:
+//
+//   {"key":"failure.measurement",
+//    "value":{"failure":"<failure>","subtest":"download"}}
+//
+// where `<failure>` is the error that occurred serialized as string. In
+// case of failure, the subtest is over and the next event to be emitted is
+// `"status.measurement_done"`.
+//
+// Otherwise, the download subtest starts and we see the following event:
+//
+//   {"key":"status.measurement_begin",
+//    "value":{"server":"<server>","subtest":"download"}}
+//
+// where `<server>` is the FQDN of the server we're using. Then there
+// are zero or more events like:
+//
+//   {"key": "measurement", "value": <value>}
+//
+// where `<value>` is a serialized spec.Measurement struct. Note that
+// the minimal `<value>` MUST contain a field name `"subtest"` with
+// value equal either to `"download"` or `"upload"`.
+//
+// Finally, this event is always emitted at the end of the subtest:
+//
+//   {"key":"status.measurement_done","value":{"subtest":"download"}}
+//
+// The upload subtest is like the download subtest, except for the
+// value of the `"subtest"` key.
+//
+// Exit code
+//
+// This tool exits with zero on success, nonzero on failure. Under
+// some severe internal error conditions, this tool with exit using
+// a nonzero exit code without being able to print a diagnostic
+// message explaining the error that occurred. In all other cases,
+// checking the logs should help to understand the error.
+package main
+
+import (
+	"context"
+	"flag"
+	"os"
+	"time"
+
+	"github.com/m-lab/ndt7-client-go"
+	"github.com/m-lab/ndt7-client-go/spec"
+)
+
+var flagBatch = flag.Bool("batch", false, "emit JSON events on stdout")
+
+var flagHostname = flag.String("hostname", "", "optional ndt7 server hostname")
+
+var flagTimeout = flag.Int64(
+	"timeout", 45, "seconds after which the ndt7 test is aborted",
+)
+
+func downloadUpload(
```

Comment:
This is a bad name. How about `runTest` ?

And shouldn't this be a method attached to the client object?  Then you could call `client.RunDownloadTest(ctx, emitter)`

Reply:
I renamed the function. With respect to attaching this as a method to the client, I'd rather not do that. The public API of client is what I consider a minimal natural Go API.

For example, in the current API it's clear when to stop and when to fail.

If I attach also this method to the client, I basically introduce a secondary API that is based on an event emitter. Now, I am not super happy about having more than one public API because that complicates the job of people understanding the package.

Also, in general, I find it simpler to have an API where errors are reported in a sync way rather than in an async way. But errors are async in this API with an emitter.

Yet, I see what you mean here. The code in main should be refactored to be more pretty so that a future refactoring that may make it public is enabled.

Hence, I'm going to fully write the main-related code in a proper way, but I'm going to make it also internal. I don't want to expose APIs that people may rely on.

(I spent some time thinking about all of this, and I enjoyed this review because it forced me to think about what to expose and how to expose it more, thanks)